parser.rst

```diff
@@ -0,0 +1,901 @@
+program in memory before parsing it but also allows the parser to backtrack
+arbitrarily. This is made efficient by memoizing the rules already matched for
+each position. The cost of the memoization cache is that the parser will naturally
+use more memory than a simple LL(1) parser, which normally are table-based. 
+
+
+Key ideas
+~~~~~~~~~
+
+.. important::
+    Don't try to reason about a PEG grammar in the same way you would to with an EBNF
+    or context free grammar. PEG is optimized to describe **how** input strings will
+    be parsed, while context-free grammars are optimized to generate strings of the
+    language they describe (in EBNF, to know if given string is in the language you need
+    to do work to find out as is not immediately obvious from the grammar).
+
+* Alternatives are ordered ( ``A | B`` is not the same as ``B | A`` ).
+* If a rule returns a failure doesn't mean that the parsing has failed,
+  it just means "try something else".
+* By default PEG parsers run in exponential time and they are keep linear By
+  using memoization.
+* If parsing fails completely (no rule can succeed parsing all the input text), the
+  PEG parser doesn't have a concept of "where the syntax error is".
+
+
+Consequences or the ordered choice operator
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. _consequences-of-ordered-choice:
+
+Although PEG may look like EBNF, its meaning is quite different. The fact
+that in PEG parsers alternatives are ordered (which is at the core of how PEG
+parsers work) has deep consequences, other than removing ambiguity.
+
+If a rule has two alternatives and the first of them succeeds, the second one is
+**not** attempted even if the caller rule fails to parse the rest of the input.
+This is often described that the parser is "eager". To illustrate this, consider
+the following two rules (in these examples, a token is an individual character): ::
+
+    first_rule:  ( 'a' | 'aa' ) 'a'
+    second_rule: ('aa' | 'a'  ) 'a'
+
+In a regular EBNF grammar, both rules specify the language ``{aa, aaa}`` but
+in PEG, one of these two rules accepts the string ``aaa`` but not the string
+``aa``. The other does the opposite -- it accepts the string the string ``aa``
+but not the string ``aaa``. The rule ``('a'|'aa')'a'`` does
+not accept ``aaa`` because ``'a'|'aa'`` consumes the first ``a``, letting the
+final ``a`` in the rule consume the second, and leaving out the third ``a``.
+As the rule has succeeded, no attempt is ever made to go back and let
+``'a'|'aa'`` try the second alternative. The expression ``('aa'|'a')'a'`` does
+not accept ``aa`` because ``'aa'|'a'`` accepts all of ``aa``, leaving nothing
+for the final ``a``. Again, the second alternative of ``'aa'|'a'`` is not
+tried.
+
+.. caution::
+
+    The effects of ordered choice such as illustrated above, may be hidden by many levels of rules.
+
+For this reason, avoid writing rules where an alternative is contained in the next one,
```

Comment:
I think that the wording here should be stronger than just "avoid writing this". Maybe something like this?

```suggestion
For this reason, writing rules where an alternative is contained in the next one is in almost all cases a mistake,
```